### DISCUSSION TITLE
How to use StreamingResponse Callbacks with Fastapi

### DISCUSSION THREAD

**Discussion by Rohith295** [ajndkr/lanarky#172] [Q&A] [answered]
Hello, I have an existing FastAPI application that integrates LangChain. I'm looking to add a new endpoint to support conversation streams. Currently, I'm using the `ChatOpenAI` class within the `ConversationalRetrievalChain`. I need guidance on setting up an endpoint that, when triggered with a user prompt, initiates streaming from this chain. Despite searching through various discussions, I haven't found examples that specifically use FastAPI with this setup. Could someone provide assistance on how to achieve this?

**Comment by ajndkr** [comment, marked as answer]
Hi @Rohith295! thanks for checking out lanarky!

To setup a streaming response with fastapi, you can check out the docs: https://lanarky.ajndkr.com/learn/adapters/langchain/fastapi/#streaming

Let me know if it helps!

**Reply by Rohith295** [reply]
Thanks @ajndkr It really saved a lot of effort.

**Reply by Rohith295** [reply]
@ajndkr one more question. After using TokenStreamHandler, the response contains both question and answer as part of the token generated. Is there a way to exclude the question as part of stream of tokens.

For example:
User Prompt: `what did Abraham lincoln do for america?`
Assitant Response(After joining the tokens): `**What did Abraham Lincoln do for America?**Abraham Lincoln fought for the equality of all human beings regardless of their race or skin color. He was assassinated for his beliefs, and his death led to an outpouring of grief and tributes from across the political spectrum in America and internationally. `

I have inherited the token handler and in method on_chain_end i am printing the output response.
`Abraham Lincoln fought for the equality of all human beings regardless of their race or skin color. He was assassinated for his beliefs, and his death led to an outpouring of grief and tributes from across the political spectrum in America and internationally.`

**Reply by ajndkr** [reply]
> After using TokenStreamHandler, the response contains both question and answer as part of the token generated. Is there a way to exclude the question as part of stream of tokens.

@Rohith295 This is very odd. The `TokenStreamHandler` simply streams the tokens incoming from the LLM. If possible, can you please share a code snippet for further investigation? (it's okay if you just share the relevant bits, full code is not needed)

**Reply by Rohith295** [reply]
```
class CustomTokenStreamingCallbackHandler(TokenStreamingCallbackHandler):
    def __init__(self, chat_session_manager: ChatSessionManager, chat_session: ChatSession, chat_history: list, *,
                 output_key: str,
                 **kwargs: dict[str, Any]):
        super().__init__(output_key=output_key, **kwargs)
        self.chat_session_manager = chat_session_manager
        self.chat_session = chat_session
        self.chat_history = chat_history

    def on_chain_end(
            self, outputs: dict[str, Any], **kwargs: dict[str, Any]
    ) -> None:
        if "answer" in outputs:
            composite_key = self.chat_session_manager._generate_composite_key(self.chat_session)
            # Insert/Update chat history
            self.chat_session_manager._update_session_history(chat_session=self.chat_session,
                                                              composite_key=composite_key,
                                                              history=self.chat_history,
                                                              assistant_response=outputs['answer'])
```

```
def get_streaming_prompt_response(self, chat_session: ChatSession) -> StreamingResponse:
        composite_key = self._generate_composite_key(chat_session)
        # Retrieve existing chat history or initialize an empty one
        current_session_chat_history = self.get_chat_history(unique_key=composite_key)

        resp = StreamingResponse(chain=self.chatbot.chain,
                                 config={"inputs": {"question": chat_session.user_prompt,
                                                    "chat_history": [(user_message, assistant_message) for
                                                                     _, user_message, assistant_message in
                                                                     current_session_chat_history] if current_session_chat_history else []},
                                         "callbacks": [
                                             CustomTokenStreamingCallbackHandler(
                                                 chat_session_manager=self,
                                                 chat_session=chat_session,
                                                 chat_history=current_session_chat_history,
                                                 output_key=self.chatbot.chain.output_key,
                                                 mode=TokenStreamMode.JSON)
                                         ]})
        return resp
```

This is how i am using it.

**Reply by ajndkr** [reply]
Thanks! I will take a look. It's possible `ConversationalRetrievalChain` needs some additional logic.

**Reply by ajndkr** [reply]
@Rohith295 there's no issue with the callback handler. I believe the root cause of this issue is the chain definition. The reason why the assistant response is combining the question and the final answer is because the `question_generator` attribute of `ConversationalRetrievalChain` is using llm with streaming enabled. You need to enable streaming only for the `combine_docs_chain` attribute.

Example:

```
from langchain.chains import ConversationalRetrievalChain
from langchain.chat_models import ChatOpenAI
from langchain.callbacks import StreamingStdOutCallbackHandler

llm = ChatOpenAI(streaming=True)
chain = ConversationalRetrievalChain.from_llm(
    llm=llm,
    retriever=vectorstore.as_retriever(),
    condense_question_llm=ChatOpenAI(),
)
```

If you don't pass the `condense_question_llm` parameter, `from_llm` classmethod uses `llm` parameter to initialise `question_generator` attribute.

let me know if this helps!

**Reply by Rohith295** [reply]
@ajndkr it works. Thanks. If you don't mind me asking how did you approach this? The langchain's documentation is so confusing. I hardly find it useful.

**Reply by ajndkr** [reply]
@Rohith295 super! 

> If you don't mind me asking how did you approach this? The langchain's documentation is so confusing. I hardly find it useful.

Yeah sure! I agree the documentation can be confusing at times. Especially for this chain type, there's no example code/notebook. There was only some code in the API reference but it wasn't super useful. In such cases when the documentation requires more work, the best option is to just dive into the source code. It's quite easy with an IDE (i use vscode). There I found that the `from_llm` classmethod reuses the same `llm` instance for all sub-chains unless you pass the additional parameter i mentioned above. 

About debugging the callback handlers, I always use langchain's `StreamingStdOutCallbackHandler` callback handler. It helps me understand if the lanarky callback handlers contain bugs or the chain. In this case, I found that the chain had the bug. 

Beyond that, I have been using Langchain since the start of 2023 so I have developed a good enough intuition about knowing where the bug(s) might be 😅

**Reply by Rohith295** [reply]
Haha. Thanks for the pointers @ajndkr .

**Reply by Rohith295** [reply]
@ajndkr Hello Again, I have another question. I have source documents as part of the vector db. How can i include them while streaming the tokens? Is there any recommended way, because i have enabled return_source_documents=True, but only see them when i am doing chain.invoke() as it returns source_documents key in the output, but in the streaming api, its only tokens which are generated by the llm.

**Reply by ajndkr** [reply]
@Rohith295 you'll need to pass the `SourceDocumentStreamingCallbackHandler` along with the token streaming callback. You can find all the callback handlers provided by lanarky here: https://lanarky.ajndkr.com/learn/adapters/langchain/callbacks/

**Reply by Rohith295** [reply]
Can you provide with sample code. I did added as part callbacks handler list, but still running into run time errors.

Issue fixed.

**Comment by ajndkr** [comment]
@Rohith295 please feel free to open new issues/discussions for new topics. I will close this discussion for now as the main question has been answered.
